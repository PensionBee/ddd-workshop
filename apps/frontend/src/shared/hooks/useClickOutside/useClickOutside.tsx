import { useEffect } from "react";

export const useClickOutside = <E extends HTMLElement>(
  ref: React.RefObject<E>,
  callback: (event: MouseEvent | TouchEvent) => void
) => {
  useEffect(() => {
    const listener = (event: MouseEvent | TouchEvent) => {
      if (!ref.current) return;
      if (event.target && ref.current.contains(event.target as Node)) return;
      return callback(event);
    };
    document.addEventListener("mousedown", listener);
    document.addEventListener("touchstart", listener);
    return () => {
      document.removeEventListener("mousedown", listener);
      document.removeEventListener("touchstart", listener);
    };
  }, [ref, callback]);
};
