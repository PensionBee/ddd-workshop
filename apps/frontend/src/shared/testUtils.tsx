import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
