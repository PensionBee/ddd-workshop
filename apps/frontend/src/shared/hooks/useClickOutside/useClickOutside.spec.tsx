import { act, fireEvent, render } from "@shared/testUtils";
import { vi } from "vitest";
import { useRef } from "react";
import { useClickOutside } from "./useClickOutside";

const MOCK_CALLBACK = vi.fn();

const MOUSE_DOWN_EVENT = new MouseEvent("mousedown", {
  bubbles: true,
  cancelable: true,
});

const TOUCH_START_EVENT = new MouseEvent("touchstart", {
  bubbles: true,
  cancelable: true,
});

const Component = () => {
  const ref = useRef(null);
  useClickOutside(ref, MOCK_CALLBACK);

  return (
    <div data-testid="parent-element">
      <div data-testid="ref-element" ref={ref}>
        <div data-testid="child-element" />
      </div>
      <div data-testid="outside-element" />
    </div>
  );
};

const NoRefComponent = () => {
  useClickOutside({} as React.MutableRefObject<null>, MOCK_CALLBACK);

  return (
    <div data-testid="parent-element">
      <div data-testid="ref-element">
        <div data-testid="child-element" />
      </div>
      <div data-testid="outside-element" />
    </div>
  );
};

describe("useClickOutside", () => {
  it("should fire callback when clicking outside the element", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const outsideElement = getByTestId("outside-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(outsideElement, MOUSE_DOWN_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should fire callback when touching outside the element", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const outsideElement = getByTestId("outside-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(outsideElement, TOUCH_START_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should fire callback when clicking on the elements parent", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const parentElement = getByTestId("parent-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(parentElement, MOUSE_DOWN_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should NOT fire callback when clicking on the element", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const refElement = getByTestId("ref-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(refElement, MOUSE_DOWN_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should NOT fire callback when clicking on the elements children", () => {
    const { asFragment, getByTestId } = render(<Component />);
    const childElement = getByTestId("child-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(childElement, MOUSE_DOWN_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not fire listener if ref is not set", () => {
    const { asFragment, getByTestId } = render(<NoRefComponent />);
    const outsideElement = getByTestId("outside-element");

    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    act(() => fireEvent(outsideElement, MOUSE_DOWN_EVENT));
    // expect(MOCK_CALLBACK.mock.calls.length).toBe(0);
    expect(asFragment()).toMatchSnapshot();
  });
});
