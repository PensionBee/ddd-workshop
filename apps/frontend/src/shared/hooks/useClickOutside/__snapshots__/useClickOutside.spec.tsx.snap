// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useClickOutside > should fire callback when clicking outside the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should NOT fire callback when clicking on the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should NOT fire callback when clicking on the elements children 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should fire callback when clicking on the elements parent 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should fire callback when clicking outside the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should fire callback when touching outside the element 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;

exports[`useClickOutside should not fire listener if ref is not set 1`] = `
<DocumentFragment>
  <div
    data-testid="parent-element"
  >
    <div
      data-testid="ref-element"
    >
      <div
        data-testid="child-element"
      />
    </div>
    <div
      data-testid="outside-element"
    />
  </div>
</DocumentFragment>
`;
